### PR TITLE
[region-isolation] Add support for transferring parameters.

### PR DESCRIPTION
**Pull request by gottesmm**
A transferring parameter is a consuming like param modifier that can be used to transfer values to an async function such that you can transfer the value further.

Since it is consuming, one can assign into it. When one assigns into it, one has transferred the new value into the transferred parameter.

### CODE REVIEW

**Comment by gottesmm** [comment]
@swift-ci smoke test

File: test/Concurrency/transfernonsendable_strong_transferring.swift [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+func testNonStrongTransferDoesntMerge() async {
+}
+
+//////////////////////////////////
+// MARK: Transferring Parameter //
+//////////////////////////////////
+
+func testTransferringParameter_canTransfer(_ x: transferring Klass, _ y: Klass) async {
+  await transferToMain(x)
+  await transferToMain(y) // expected-warning {{call site passes `self` or a non-sendable argument of this function to another thread, potentially yielding a race with the caller}}
```
Comment (ktoso) [inline test/Concurrency/transfernonsendable_strong_transferring.swift:69]: I'm wondering if the warning should be phrased in terms of actor boundary crossing and not "to another thread", or is this intentional?

Comment (gottesmm) [reply test/Concurrency/transfernonsendable_strong_transferring.swift:69]: I agree that said diagnostic needs to be wordsmithed (and actually split up into 1-2 other ones), this PR is just bringing up transferring, not solving that problem.

Comment (ktoso) [reply test/Concurrency/transfernonsendable_strong_transferring.swift:69]: Yea not a blocker for merging ofc, just something I noticed

**Comment by gottesmm** [comment]
apple/swift-syntax#2410

@swift-ci smoke test

File: test/Concurrency/transfernonsendable_strong_transferring.swift (hunk 1 above)
Comment (gottesmm) [reply test/Concurrency/transfernonsendable_strong_transferring.swift:69]: Thanks for bringing it up = ).

**Comment by gottesmm** [comment]
https://github.com/apple/swift-syntax/pull/2410

@swift-ci smoke test

**Comment by gottesmm** [comment]
(just rebasing/retesting since I am back from vacation)

**Comment by gottesmm** [comment]
Looks like the windows/macOS failures are related to swiftpm

**Comment by gottesmm** [comment]
I got feedback from some people about some small changes about loosening the rules around assigning into a transferring parameter. I am going to do that in a follow on commit. This at least gets in the basics and ensures that I do not hit more merge conflicts around Serialization.

**Pull request merged**